operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu

```diff
@@ -0,0 +1,160 @@
+__constant__ double d_vlp16_cos_pitch[data_collection::sensors::kVLP16LineCount];
+
+// Quick return cuda error and error message. Will force quit program cause all
+// cuda errors are fatal errors.
+#define CHECK_CUDA_ERROR(err) (HandleError(err, __FILE__, __LINE__))
+
+static void HandleError(cudaError_t err, const char* file, int line) {
+  if (err != cudaSuccess) {
+    fprintf(stderr,
+            "Error %d: \"%s\" in %s at line %d\n",
+            int(err),
+            cudaGetErrorString(err),
+            file,
+            line);
+    exit(int(err));
+  }
+}
+
+namespace data_collection {
+namespace sensors {
+
+__device__ uint16_t ConvertSmallEndianToBigEndian(uint16_t small_endian) {
+  const uint8_t* bits = reinterpret_cast<const uint8_t*>(&small_endian);
+  return (bits[1] << 8) + bits[0];
+}
+
+/// @brief  Convert a raw Velodyne VLP-16 packet to a list of XYZ points.
+/// @param d_packet The 1206-byte packet in device memory to convert.
+/// @param d_xyzintensity_list The device memory destination for 384 XYZ points.
+/// @param d_cos_rot_table Precomputed values for cosines of azimuth angles.
+/// @param d_sin_rot_table Precomputed values for sines of azimuth angles.
+/// @note The VLP-16 firing sequence does not follow the physical order of lasers,
+///       but instead "jumps around". See the Velodyne VLP-16 User Manual for details.
+__global__ void ConvertRawPacketToXYZ(data_collection::sensors::RawVelodynePacket* d_packet,
+                                      Blocks* d_xyzintensity_list, double* d_cos_rot_table,
+                                      double* d_sin_rot_table) {
+  // x: 0 -- 31, y: 0 -- 11
+  // i: 0 -- 31, j: 0 -- 11
+  int i = threadIdx.x;
+  int j = threadIdx.y;
+
+  // out of size
+  if (i >= kVelodyneRecords || j >= kVelodyneBlocks) { return; }
+
+  // Convert from the default Velodyne little endian format to the default
+  // IGX big endian format.
+  uint16_t azimuth =
+      ConvertSmallEndianToBigEndian(d_packet->blocks_[j].azimuth_hundredths_degrees_);
+  float range_meters =
+      ConvertSmallEndianToBigEndian(d_packet->blocks_[j].records_[i].distance_two_millimeters_) *
+      kRawVelodyneDefaultDistanceAccuracy * kMillimetersToMeters;
+
+  // Use pre-computed lookup tables to translate from spherical coordinates
+  // to Cartesian coordinates.
+  d_xyzintensity_list->blocks_[j].lines_[i].x =
+      range_meters * d_vlp16_cos_pitch[i] * d_sin_rot_table[azimuth];
+  d_xyzintensity_list->blocks_[j].lines_[i].y =
+      range_meters * d_vlp16_cos_pitch[i] * d_cos_rot_table[azimuth];
+  d_xyzintensity_list->blocks_[j].lines_[i].z = range_meters * d_vlp16_sin_pitch[i];
+}
+
+/// @brief Pre-compute sine and cosine values for rapid lookup.
+void VelodyneConvertXYZHelper::InitSinAndCosTable() {
+  HOLOSCAN_LOG_DEBUG("Initializing sin and cos tables in GPU.");
+  size_t size_of_rot_table = kVelodyneMaxAzimuth * sizeof(double);
+  sin_rot_table_ = (double*)malloc(size_of_rot_table);
```

Comment:
The CPU memory for the sin and cos tables is not used outside of this function. You might want to have a local variable for this. Also using `std::unique_ptr` should simplify things.

Reply:
Thanks, good catch. Fixed with local `std::vector` for array memory management.